File: src/bin/day09.rs
use std::fs;

fn solve(nums: &[i64], step: usize) -> Option<i64> {
    let mut start = 0;
    while start + step < nums.len() {
        let slice = &nums[start..start + step];
        let last = nums[start + step];
        let mut found = false;
        for a in slice {
            for b in slice {
                if a == b {
                    continue;
                }
                if last == a + b {
                    found = true;
                }
            }
        }
        if !found {
            return Some(last);
        }
        start += 1;
    }
    None
}

fn solve2(nums: &[i64], target: i64) -> Option<i64> {
    for start in 0..nums.len() {
        for end in start+1..nums.len() {
            let slice = &nums[start..end];
            let sum = slice.iter().sum::<i64>();
            if sum > target {
                break;
            }
            if sum == target {
                let min = slice.iter().min().unwrap();
                let max = slice.iter().max().unwrap();
                return Some(min + max);
            }
        }
    }
None
}

fn main() {
    let contents = fs::read_to_string("inputs/day09.txt").expect("Something went wrong reading the file");
    let nums: Vec<i64> = contents
        .lines()
        .map(|s| s.parse::<i64>().unwrap())
        .collect();

    if let Some(target) = solve(&nums, 25) {
        println!("{:?}", target);
        println!("{:?}", solve2(&nums, target));
    }

    println!("done");
}

File: julia/day09.jl
nums = open("inputs/day09.txt") do f
    parse.(Int64, readlines(f))
end

function solve(nums, step=25)::Union{Int64,Nothing}
    start = 1
    for i in start + step:length(nums)
        slice = view(nums, i:(i + step))
        last = nums[i + step]
        search = function ()
            for a in slice
                for b in slice
                    if last == a + b
                        return true
                    end
                end
            end
            return false
        end
        if !search()
            return last
        end
    end
end


function solve2(nums, target)::Union{Int64,Nothing}
    for start in 1:length(nums)
        mx = nums[start]
        mn = nums[start]
        acc = nums[start]
        for i in (start + 1):length(nums)
            acc += nums[i]
            mx = max(mx, nums[i])
            mn = min(mn, nums[i])
            if acc > target
                break
            end
            if acc == target
                return mn + mx
            end
        end
    end
end

function main()
    part1 = solve(nums)
    println("Part 1: ", part1)
    println("Part 2: ", solve2(nums, part1))
end

main()